### PR TITLE
[merged] Atomic/atomic.py: Deal with non-ascii input

### PR DESCRIPTION
**Pull request by baude**
https://bugzilla.redhat.com/show_bug.cgi?id=1323165#c2 suggests
that atomic install does not deal with non-ascii input very
well.  I believe this addresses the concerns with the proper
UTF-8 encode and decode operations.

### CODE REVIEW

**Comment by rhatdan** [comment]
@adelton  PTAL

**Comment by cgwalters** [comment]
I'm almost certain this is going to need to handle Python 2 vs Python 3.  Did you test it in both scenarios?

http://lucumr.pocoo.org/2014/5/12/everything-about-unicode/

**Comment by baude** [comment]
@cgwalters Thanks for the reference, Im going to sit elsewhere and read it.  I think I have the conditional handled but maybe the article will inform me differently.

File: Atomic/util.py [hunk 1]
```diff
@@ -119,7 +123,7 @@ def default_ro_container_context():
 
 def write_out(output, lf="\n"):
     sys.stdout.flush()
-    sys.stdout.write(str(output) + lf)
+    sys.stdout.write(output + lf)
```
Comment (cgwalters) [inline Atomic/util.py]: For example it looks like the `str` there came from 6e1f4dde ...which seems like it was probably copied from somewhere for a reason.  I'm guessing it's something like needing to handle stdout being passed unicode but expecting bytes?

**Comment by baude** [comment]
Passes all tests with python2 and python3 including your proposed new test -> https://github.com/projectatomic/atomic/pull/358

File: Atomic/util.py [hunk 2]
```diff
@@ -119,7 +123,7 @@ def default_ro_container_context():
 
 def write_out(output, lf="\n"):
     sys.stdout.flush()
-    sys.stdout.write(str(output) + lf)
+    sys.stdout.write(output.encode('utf-8') + lf)
```
Comment (cgwalters) [inline Atomic/util.py]: Are you really sure this works on Python 3?

```
[walters@9845f984f09d ostree]$ python3
Python 3.4.3 (default, Mar 31 2016, 20:42:37) 
[GCC 5.3.1 20151207 (Red Hat 5.3.1-2)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> "foo".encode('utf-8') + "\n"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can't concat bytes to str
>>> 
```

i.e. don't we have to do `lf.encode('utf-8')` too?

File: Atomic/atomic.py [hunk 3]
```diff
@@ -142,6 +142,9 @@ def __init__(self):
         self.atomic_config = None
         self.docker_cmd = None
         self.debug = False
+        self.is_python2 = True
+        if int(sys.version[0]) > 2:
+            self.is_python2 = False
```
Comment (rhatdan) [inline Atomic/atomic.py]: self.is_python2 = (int(sys.version[0]) <= 2)

File: Atomic/atomic.py [hunk 4]
```diff
@@ -677,7 +680,10 @@ def gen_cmd(self, cargs):
                 args.append(self.name)
                 continue
             args.append(c)
-        return " ".join(args)
+        if self.is_python2:
+            return " ".join([x.decode('utf-8') for x in args])
+        else:
+            return " ".join(x for x in args)
```
Comment (rhatdan) [inline Atomic/atomic.py]: return " ".join(args)

File: Atomic/util.py [hunk 5]
```diff
@@ -119,7 +125,10 @@ def default_ro_container_context():
 
 def write_out(output, lf="\n"):
     sys.stdout.flush()
-    sys.stdout.write(str(output) + lf)
+    if is_python2:
+        sys.stdout.write(output.encode('utf-8') + lf)
+    else:
+        sys.stdout.write(output + str(lf))
```
Comment (rhatdan) [inline Atomic/util.py]: Why is the str(lf) necessary?

File: Atomic/util.py [hunk 6]
```diff
@@ -86,9 +88,13 @@ def subp(cmd, cwd=None):
 # since we are a privileged process, and we don't want to leak things like
 # the docker socket into child processes by default
 def check_call(cmd, env=os.environ, stdin=None, stderr=None, stdout=None):
-    # Make sure cmd is a list
+    # Make sure cmd is a list; break if needed
     if not isinstance(cmd, list):
-        cmd = shlex.split(cmd)
+        try:
```
Comment (rhatdan) [inline Atomic/util.py]: Should we just do the is_python2 check here?

File: Atomic/util.py (hunk 5 above)
Comment (baude) [inline Atomic/util.py]: BC in python 3, output is a str and lf is a byte and then cannot be concatenated.

File: Atomic/util.py (hunk 6 above)
Comment (baude) [inline Atomic/util.py]: Yeah, seems prudent.

File: Atomic/util.py (hunk 5 above)
Comment (rhatdan) [inline Atomic/util.py]: ok.

File: Atomic/util.py [hunk 7]
```diff
@@ -86,9 +88,13 @@ def subp(cmd, cwd=None):
 # since we are a privileged process, and we don't want to leak things like
 # the docker socket into child processes by default
 def check_call(cmd, env=os.environ, stdin=None, stderr=None, stdout=None):
-    # Make sure cmd is a list
+    # Make sure cmd is a list; break if needed
     if not isinstance(cmd, list):
-        cmd = shlex.split(cmd)
+        if not is_python2:
```
Comment (rhatdan) [inline Atomic/util.py]: Can you make this positive like you did below.

Comment (rhatdan) [inline Atomic/util.py]:
```
if is_python2:
            # The command contains a non-ascii character
            cmd = shlex.split(" ".join([x.encode('utf-8') for x in cmd.split()]))
else:
...
```

**Comment by rhatdan** [comment]
@cgwalters-bot r+

**Pull request closed without merging**